src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -137,17 +132,33 @@ public ICollection<string> Keys
         {
             get
             {
-                EnsureWritable();
-                return InnerDictionary.Keys;
+                Upgrade();
```

Comment:
There's little payoff to trying to produce optimized implementations of everything on `IDictionary<,>` and `ICollection<>`. All of these extra implementations require their own tests.

The eventual goal of all of this is that an app using a 'normal' MVC action selection scheme or attribute routing doesn't need to `Upgrade` on any of the main code paths.